### PR TITLE
Loan Process - 2nd version

### LINKED ISSUE #232: Loan Process - 2nd version
My initial perceptions were that a Loan was for one User for one Batch.

But a Loan is for all items lent from a batch.

So, `Loan` doesn't actually store any unique data - it's just a list of items from a batch.  

We can _actually_ support that by just having a `loanedTo` field in `item`:
```
loanedTo: User['id''] | undefined
```

Note: this strategy removes the `Loan` and `LoanItem` tables, and makes it **much** easier find items loaned to a user.

So, when a group of items are loaned to a user, we just put that userId in `loanedTo`.

When items are returned, we just clear the `loanedTo` flag.

For the `Loan` process we have a remark field.  If it is provided, it will get included in the audit entry.

We need a bit of an audit trail to cover this, so for the loan, we create an audit entry for:
```
activity: `ITEM_LOAN`
comment: `Item loaned to ${user} by ${currentUser}. ${remark}`
id: itemId
```

```
activity: `ITEM_RETURN`
comment: `Item returned to ${user} by ${currentUser}. ${remark}`
id: itemId
```

Note: for the above we need to introduce an `id` field for `trackEvent`.  This will be eventually id for a number of table types, but we know which from the `activity` type.

Note 2: the `item` will still have a value in `vaultLocation`. The users actually want this - so that when the item is returned, they know which shelf to put it back on.  But, we'll have to modify the `Muster Reports` to only include items that are not on loan.

Note 3: with this new data model in place we can add `Loaned to` to the `ItemList` filters and columns (hidden by default).

### PR DESCRIPTION
**Pull request by salony75way**
fixes #232

### CODE REVIEW

**Comment by IanMayo** [comment]
Compile error to be fixed, please @salony75way :

Error: src/providers/dataProvider/index.ts(83,11): error TS2339: Property 'name' does not exist on type 'true | { name: string; }'.

**Comment by IanMayo** [comment]
Hello @salony75way - the architecture still isn't quite right. We no longer require the `Loan` entity.  We just have some loan-related attributes to `Item`.   Updated schema [here](https://docs.google.com/drawings/d/1SYIb0PZxcabvhXdNMAEHlbdWaoam2ipB7x5ZUudVVGc/edit)

**Comment by salony75way** [comment]
> Hello @salony75way - the architecture still isn't quite right. We no longer require the `Loan` entity. We just have some loan-related attributes to `Item`. Updated schema [here](https://docs.google.com/drawings/d/1SYIb0PZxcabvhXdNMAEHlbdWaoam2ipB7x5ZUudVVGc/edit)

Hi @IanMayo , Yes i have removed all resources related to loan. Let me know if we need to remove some other dependency in code.

**Comment by IanMayo** [comment]
> Hi @IanMayo , Yes i have removed all resources related to loan. 

We still have the `Loan` type, and we still refer to it from `LoanForm`:
image

**Comment by IanMayo** [comment]
One other change please. It's about `loanedBy`.  This suggests it is the librarian that lent the book. This should be `loanedTo` - so it's the person that has taken the book away.

**Comment by IanMayo** [comment]
Hello @salony75way - IMHO the `loan` form and buttons should be moved to the `Items` resource - since they are only used for that resource.

**Comment by IanMayo** [comment]
Aah, one more thing. Can we have a `notify` popup once the database changes are complete?

File: src/providers/dataProvider/index.ts [hunk 1]
```diff
@@ -78,33 +78,76 @@ const withCreatedBy = (
...
+        data: { name }
+      } = await provider.getOne<User>(constants.R_USERS, {
+        id: holder
+      })
+
+      const promisees = items.map(async (item) => {
+        await audit({
+          type: AuditType.ITEM_LOAN,
+          activityDetail: `Item loaned to ${name} by ${userName}. ${remarks}`,
+          resource: constants.R_LOAN_ITEMS,
```
Comment (IanMayo) [inline src/providers/dataProvider/index.ts:106]: This should be `constants.R_ITEMS`

We should probably delete `constants.R_LOAN_ITEMS`

File: src/types.d.ts [hunk 2]
```diff
@@ -108,6 +103,7 @@ interface Item {
   musterRemarks: string
   createdAt: string
   createdBy: User['id']
+  loanedBy?: number
```
Comment (IanMayo) [inline src/types.d.ts:106]: this should be `loanedTo`

**Comment by IanMayo** [comment]
Have just been talking to users. They have said they don't need the `Remarks` field in the `LoanForm`.   It's not on the legacy system, nor is it in the system their sister-organisation uses.

So, it will get deleted from the `remarks` too, please.

**Pull request merged**